gdist_c_api.cpp

```diff
@@ -0,0 +1,74 @@
+#include <iostream>
+#include <fstream>
+
+#include "geodesic_library/geodesic_algorithm_exact.h"
+
+
+double* compute_gdist_impl(
+    unsigned number_of_vertices,
+    unsigned number_of_triangles,
+    double *vertices,
+    int *triangles,
+    unsigned number_of_source_indices,
+    unsigned number_of_target_indices,
+    unsigned *source_indices_array,
+    unsigned *target_indices_array,
+    double distance_limit = geodesic::GEODESIC_INF
+) {
+
+    std::vector<double> points (vertices, vertices + number_of_vertices);
+    std::vector<unsigned> faces (triangles, triangles + number_of_triangles);
+    std::vector<unsigned> source_indices (source_indices_array, source_indices_array + number_of_source_indices);
+    std::vector<unsigned> target_indices (target_indices_array, target_indices_array + number_of_target_indices);
+
+    geodesic::Mesh mesh;
+    mesh.initialize_mesh_data(points, faces); // create internal mesh data structure including edges
+
+    geodesic::GeodesicAlgorithmExact algorithm(&mesh); // create exact algorithm for the mesh
+    
+    std::vector<geodesic::SurfacePoint> all_sources, stop_points;
+
+    for (unsigned index: source_indices) {
+        all_sources.push_back(geodesic::SurfacePoint(&mesh.vertices()[index]));
+    }
+
+    for (unsigned index: target_indices) {
+        stop_points.push_back(geodesic::SurfacePoint(&mesh.vertices()[index]));
+    }
+
+    algorithm.propagate(all_sources, distance_limit, &stop_points);
+
+    // double distance[target_indices.size()];
+    double *distance = (double *)malloc(sizeof(double) * target_indices.size());
+    for (int i = 0; i < (int)stop_points.size(); ++i) {
+        algorithm.best_source(stop_points[i], distance[i]);
```

Comment:
This will be fine for small cases, but in real data, this will be a large sparse matrix you are returning as a dense matrix, requiring a lot more memory than it should.

I would suggest doing this source by source point, and constructing the arrays for a CSR sparse matrix, which you can then return in the Python API.

Also remember to that your allocation here isn't being freed anywhere, so it would be preferable to allocate in Python via numpy and pass the buffer to this function so it can write results to the buffer.  If the user's Gdist object is GC'd, Python will clean up memory for you.  This would also align with the goal of using a CSR representation, since you'll need the C/C++ code to fill in 3 arrays instead of 1.

Reply:
This code is for `compute_gdist`. While using `local_gdist_matrix` we are using the sparse matrix. This is not implemented yet.

In the `local_gdist_matrix` case I have one query.  How we can know the number of elements in `row`, `column`, and `data` beforehand in order to allocate memory?

Reply:
ok I didn't have that context here to work with. 

As for allocation, you can work row by row: allocate memory for a full row, get the gdist values, see where the values are non-zero, push back non-zero elements to a vector and keep going.

Reply:
This has the downside of creating mesh every time for each row. But, I guess that's the only way to do where we can allocate memory on the Python side.

Reply:
If this is a lower level function being called elsewhere, that's true.  Maybe we can come back to this in a second PR, because it'd be more important to move on to building and packaging, than to have it return a sparse matrix, which can be done in a second step, time permitting.